tests/Sphere3d.elm

```diff
@@ -0,0 +1,191 @@
+                Sphere3d.centerPoint sphere
+                    |> Point3d.squaredDistanceFromAxis axis
+                    |> Expect.within (Absolute 1.0e-6) 0
+            )
+        ]
+
+
+throughPointsManual : Test
+throughPointsManual =
+    let
+        testCases =
+            [ ( ( ( 1, 0, 0 ), ( 0, 1, 0 ), ( -1, 0, 0 ), ( 0, 0, 1 ) )
+              , ( ( 0, 0, 0 ), 1 )
+              )
+            , ( ( ( 1, 0, 0 ), ( 0, 1, 0 ), ( -1, 0, 0 ), ( 0, 0, -1 ) )
+              , ( ( 0, 0, 0 ), 1 )
+              )
+            , ( ( ( 1, 0, 0 ), ( 0, 1, 0 ), ( -1, 0, 0 ), ( 0, 0, 0.5 ) )
+              , ( ( 0, 0, -0.75 ), 1.25 )
+              )
+            , ( ( ( 1, 0, 0 ), ( 0, 1, 0 ), ( -1, 0, 0 ), ( 0, 0, -0.5 ) )
+              , ( ( 0, 0, 0.75 ), 1.25 )
+              )
+            ]
+
+        input ( p1, p2, p3, p4 ) =
+            ( Point3d.fromCoordinates p1
+            , Point3d.fromCoordinates p2
+            , Point3d.fromCoordinates p3
+            , Point3d.fromCoordinates p4
+            )
+
+        with ( centerPoint, radius ) =
+            Sphere3d.with { centerPoint = Point3d.fromCoordinates centerPoint, radius = radius }
+    in
+    Test.concat <|
+        List.map
+            (\( inputPoints, expectedSphere ) ->
+                test
+                    ("Given " ++ toString (input inputPoints) ++ " throughPoints returns " ++ toString (with expectedSphere))
+                    (\_ ->
+                        Sphere3d.throughPoints (input inputPoints)
+                            |> Maybe.map (Expect.equal (with expectedSphere))
+                            |> Maybe.withDefault (Expect.fail "throughPoints returned Nothing on valid input")
+                    )
+            )
+            testCases
+
+
+throughPointsFuzz : Test
+throughPointsFuzz =
+    fuzz4 point3d
+        point3d
+        point3d
+        point3d
+        "All given points lie on the sphere constructed using `throughPoints`"
+        (\p1 p2 p3 p4 ->
+            let
+                -- if the first three points are collinear it is not possible to construct a circle
+                -- that passes through them
+                notCollinear =
+                    Direction3d.from p1 p2
+                        |> Maybe.map
+                            (\direction ->
+                                Axis3d.with { originPoint = p1, direction = direction }
+                            )
+                        |> Maybe.map
+                            (\axis ->
+                                Point3d.squaredDistanceFromAxis axis p3 /= 0
+                            )
+
+                -- it is not possible to construct a sphere from four points that are all in the same plane
+                notInTheSamePlane =
+                    Plane3d.throughPoints ( p1, p2, p3 )
+                        |> Maybe.map (flip Point3d.signedDistanceFrom p4)
+                        |> Maybe.map ((/=) 0)
+
+                isValidInput =
```

Comment:
Similarly here I think you should be able to do a tetrahedron volume check - there's no `Tetrahedron3d` type but you can still calculate the volume with something like

```elm
v2 = Vector3d.from p1 p2
v3 = Vector3d.from p1 p3
v4 = Vector3d.from p1 p4
volume = abs (Vector3d.dotProduct v4 (Vector3d.crossProduct v2 v3))
```